src/tensors/cpu/prod_blas.h

```diff
@@ -1,13 +1,123 @@
+#pragma once
 #if MKL_FOUND
-#include <mkl.h>
+    #include <mkl.h>
 #elif BLAS_FOUND
-  #if WASM_COMPATIBLE_BLAS
-    #include "3rd_party/onnxjs/src/wasm-ops/gemm.h"
-  #else
     #include <cblas.h>
-  #endif // WASM_COMPATIBLE_BLAS
+#elif USE_ONNX_SGEMM
+    #include "3rd_party/onnxjs/src/wasm-ops/gemm.h"
+#elif USE_RUY_SGEMM
+    #include "ruy/ruy.h"
+    #include "ruy/system_aligned_alloc.h"
 #endif
 
+#if USE_RUY_SGEMM
+
+// AlignedVector allocates aligned memory and cleans up after itself. RAII
+// wrapper similar to intgemm's AlignedVector.
+template <class T>
+class AlignedVector {
+public:
+  AlignedVector(size_t num_elem)
+      : size_(num_elem),
+        storage_(reinterpret_cast<T *>(ruy::detail::SystemAlignedAlloc(sizeof(T) * num_elem))) {}
+
+  T *begin() { return storage_; }
+  T *data() { return storage_; }
+  size_t size() const { return size_; }
+  size_t memSize() const { return sizeof(T) * size_; }
+
+  // Forbid copy
+  AlignedVector(const AlignedVector &) = delete;
+  AlignedVector &operator=(const AlignedVector &) = delete;
+
+  ~AlignedVector() { ruy::detail::SystemAlignedFree(reinterpret_cast<void *>(storage_)); }
+
+private:
+  size_t size_;
+  T *storage_;
+};
+
+
+inline void GemmRuy(const bool transA,
+                    const bool transB,
+                    const int M,
+                    const int N,
+                    const int K,
+                    const float alpha,
+                    const float *A,
+                    const int lda,
+                    const float *B,
+                    const int ldb,
+                    const float beta,
+                    float *C,
+                    const int ldc) {
+  ruy::Context context;
+
+  // If we need to transpose, we can swap dimensions in layout claim the matrix
+  // is just column-major. Set ordering so transpose.
+  const auto orderA = (transA ? ruy::Order::kColMajor : ruy::Order::kRowMajor);
+  const auto orderB = (transB ? ruy::Order::kColMajor : ruy::Order::kRowMajor);
+
+  ruy::Matrix<float> lhs;
+  ruy::MakeSimpleLayout(M, K, orderA, lhs.mutable_layout());
+  lhs.set_data(A);
```

Comment:
This is zero copy, right, just pointer set?

Reply:
If you mean this statement, yes. It's a pointer set. However, I think ruy might be allocating/deallocating under the hood based on it's needs of a layout change? https://github.com/google/ruy/blob/38a9266b832767a3f535a74a9e0cf39f7892e594/ruy/prepare_packed_matrices.cc#L69-L92